sdks/python/apache_beam/transforms/combiners.py

```diff
@@ -442,58 +415,86 @@ def display_data(self):
     else:
-      lt = self._compare
+      assert not holds_comparables
 
-    threshold, buffer = accumulator
-    element_key = self._key_fn(element) if self._key_fn else element
+    comparable = (
+        cy_combiners.ComparableValue(element, self._less_than, self._key)
+        if holds_comparables else element)
 
-    if len(buffer) < self._n:
-      if not buffer:
-        return element_key, [element]
-      buffer.append(element)
-      if lt(element_key, threshold):  # element_key < threshold
-        return element_key, buffer
-      else:
-        return accumulator  # with mutated buffer
-    elif lt(threshold, element_key):  # threshold < element_key
-      buffer.append(element)
-      if len(buffer) < self._buffer_size:
-        return accumulator  # with mutated buffer
-      else:
-        self._sort_buffer(buffer, lt)
-        min_element = buffer[-self._n]
-        threshold = self._key_fn(min_element) if self._key_fn else min_element
-        return threshold, buffer[-self._n:]
+    if len(heap) < self._n:
+      heapq.heappush(heap, comparable)
     else:
-      return accumulator
+      heapq.heappushpop(heap, comparable)
+    return (holds_comparables, heap)
 
   def merge_accumulators(self, accumulators, *args, **kwargs):
     if args or kwargs:
+      self._less_than = lambda a, b: self._compare(a, b, *args, **kwargs)
       add_input = lambda accumulator, element: self.add_input(
           accumulator, element, *args, **kwargs)
     else:
+      self._less_than = self._compare
       add_input = self.add_input
 
-    total_accumulator = None
+    result_heap = None
+    holds_comparables = None
     for accumulator in accumulators:
-      if total_accumulator is None:
-        total_accumulator = accumulator
+      holds_comparables, heap = accumulator
+      if self._less_than is not operator.lt or self._key:
+        if not holds_comparables:
+          heap = [
+              cy_combiners.ComparableValue(value, self._less_than, self._key)
+              for value in heap]
+          holds_comparables = True
       else:
-        for element in accumulator[1]:
-          total_accumulator = add_input(total_accumulator, element)
-    return total_accumulator
+        assert not holds_comparables
+
+      if result_heap is None:
+        result_heap = heap
+      else:
+        for comparable in heap:
+          _, result_heap = add_input(
+              (holds_comparables, result_heap),
+              comparable.value if holds_comparables else comparable)
+
+    assert result_heap is not None and holds_comparables is not None
+    return (holds_comparables, result_heap)
+
+  def compact(self, accumulator, *args, **kwargs):
+    holds_comparables, heap = accumulator
+    # Unwrap to avoid serialization via pickle.
+    if holds_comparables:
+      return (False, [comparable.value for comparable in heap])
+    else:
+      return accumulator
 
   def compact(self, accumulator, *args, **kwargs):
```

Comment:
Let's remove this "old" compact implementation since the "new" one above supercedes it.